horovod/run/elastic/driver.py

```diff
@@ -0,0 +1,299 @@
+        self._results = {}
+        self._worker_threads = queue.Queue()
+
+    def expect(self, worker_thread):
+        self._worker_threads.put(worker_thread)
+
+    def add_result(self, key, value):
+        if key in self._results:
+            return
+        self._results[key] = value
+
+    def get_results(self):
+        while not self._worker_threads.empty():
+            worker_thread = self._worker_threads.get()
+            worker_thread.join()
+        return self._results
+
+
+class ElasticDriver(object):
+    def __init__(self, rendezvous, discovery, min_np, max_np, timeout=None, verbose=0):
+        self._rendezvous = rendezvous
+        self._host_manager = HostManager(discovery)
+        self._min_np = min_np
+        self._max_np = max_np
+        self._verbose = verbose
+
+        self._host_assignments = {}
+        self._rank_assignments = {}
+        self._world_size = 0
+
+        self._wait_hosts_cond = threading.Condition()
+        self._timeout = timeout or int(os.getenv('HOROVOD_ELASTIC_TIMEOUT', ELASTIC_TIMEOUT_SECS))
+
+        self._create_worker_fn = None
+        self._worker_clients = {}
+
+        self._worker_registry = WorkerStateRegistry(self, self._host_manager)
+        self._results = Results()
+        self._shutdown = threading.Event()
+
+        self._discovery_thread = threading.Thread(target=self._discover_hosts)
+        self._discovery_thread.daemon = True
+        self._discovery_thread.start()
+
+    def start(self, np, create_worker_fn):
+        self._create_worker_fn = create_worker_fn
+        self._activate_workers(np)
+
+    def resume(self):
+        self._activate_workers(self._min_np)
+
+    def stop(self):
+        self._shutdown.set()
+        self._discovery_thread.join()
+
+    def finished(self):
+        return self._shutdown.is_set()
+
+    def get_results(self):
+        return self._results.get_results()
+
+    def register_worker_server(self, host, slot, addresses, secret_key):
+        self._worker_clients[(host, slot)] = WorkerNotificationClient(
+            addresses, secret_key, self._verbose)
+
+    def get_worker_client(self, slot_info):
+        return self._worker_clients.get((slot_info.hostname, slot_info.local_rank))
+
+    def record_ready(self, host, slot):
+        self._worker_registry.record_ready(host, slot)
+
+    def world_size(self):
+        return self._world_size
+
+    def local_size(self, host):
+        return len(self._host_assignments[host])
+
+    def get_slot_info(self, host, slot):
+        return self._host_assignments[host][slot] if self.has_rank_assignment(host, slot) \
+            else hosts.INVALID_SLOT_INFO
+
+    def get_coordinator_info(self):
+        return self._rank_assignments.get(0)
+
+    def has_rank_assignment(self, host, slot):
+        if self._host_manager.is_blacklisted(host):
+            return False
+        return host in self._host_assignments and len(self._host_assignments[host]) > slot
+
+    @property
```

Comment:
Why a `@property`, why not `world_size`, `coordinator_info`, `results` and `finished` as well? In any case, should not return the mutable dict `self._host_assignments` but a copy of it, deep copy not necessary here